### PR TITLE
Added RedisQueue using Jedis for Redis as a queue provider

### PR DESCRIPTION
**Pull request by maheshkelkar**
Added spock tests, Jedis client for Redis client, Jedis pool for multi-threading support, Added new queue configurations (new struct), updated docs

References #3

### CODE REVIEW

File: src/main/groovy/com/github/lookout/whoas/HookRequest.groovy [hunk 1]
```diff
@@ -3,12 +3,23 @@ package com.github.lookout.whoas
 import com.fasterxml.jackson.annotation.JsonProperty
 import org.joda.time.DateTime
 
+
 class HookRequest {
```
Comment (GiriDandu) [inline src/main/groovy/com/github/lookout/whoas/HookRequest.groovy]: Can you create another class say DistributedHookRequest from this and add this parameters, because it doesn't make sense for InMemoryQueue

File: src/main/groovy/com/github/lookout/whoas/AbstractHookQueue.groovy [hunk 2]
```diff
@@ -7,6 +7,23 @@ package com.github.lookout.whoas
  * This allows for different queueing implementations behind whoas
  */
 abstract class AbstractHookQueue {
+    protected Boolean started = false
```
Comment (GiriDandu) [inline src/main/groovy/com/github/lookout/whoas/AbstractHookQueue.groovy]: What is the use case for this parameters for InMemoryQueue. Why can't this be part of RedisQueue

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+        /* Tests whether connection is dead when returning a connection
+         * to the pool
+         * */
+        poolConfig.setTestOnReturn(true);
+
+        /* Number of connections to Redis that just sit there and do
+         * nothing
+         * */
+        poolConfig.setMaxIdle(5);
```
Comment (GiriDandu) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: Can we make RedisClientFactory() configurable through deploydb. That will be useful for us to modify during run time

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        redisClientFactory = RedisClientFactory.getInstance()
+        this.hostname = queueConfig.hostname
+        this.port = queueConfig.port
+    }
+
+    /**
+     * Default constructor
+     */
+    RedisQueue() {
+        redisClientFactory = RedisClientFactory.getInstance()
```
Comment (GiriDandu) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: Hmm, you want to set (unit testing) values in default constructor ?

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    }
+    /**
+     * Return the number of elements in the queue
+     */
+    Long getSize() {
+        if (!this.started) {
+            throw new Exception("Queue must be started before this operation is invoked")
+        }
+        Jedis redisclient = null
+        Long queueSize = 0
```
Comment (GiriDandu) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: Is 0 a valid value? If so, queueSize should be initialized to -1, because of exception handling

File: src/test/groovy/com/github/lookout/whoas/RedisQueueSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        q.getSize()
+
+        then:
+        thrown Exception
+    }
+
+    def "getSize() should return 0 by default"() {
+        given:
+        RedisQueue queue = new RedisQueue()
+        //Jedis redisClient = new MockJedis("test");
```
Comment (GiriDandu) [inline src/test/groovy/com/github/lookout/whoas/RedisQueueSpec.groovy]: Remove the comment

File: src/main/groovy/com/github/lookout/whoas/HookRequest.groovy (hunk 1 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/HookRequest.groovy]: id parameter is only for debugging. I thought it will be useful in tracking/tracing the hook requests. If its not deemed necessary, I can take it out

File: src/main/groovy/com/github/lookout/whoas/AbstractHookQueue.groovy (hunk 2 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/AbstractHookQueue.groovy]: To protect the queue being used without starting(which can be used to setup the queue paramters)

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy (hunk 3 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: I dont think we want to expose it right now. Lets wait until we learn more about RedisClientFactory (i.e. JedisPool)

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 4 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: I added this to avoid code duplication.

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+        if (!this.started) {
+            throw new Exception("Queue must be started before this operation is invoked")
+        }
+
+        List<String> messages = null
+        Jedis redisclient = null
+        try {
+            redisclient = redisClientFactory.getJedis()
+            messages = redisclient.blpop(0, "queue");
```
Comment (GiriDandu) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: its a single message or multiple messages

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 5 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: We are handling the exception in this function. So we are returning "0" in case of errors. Instead of returning -1, we should throw exception, but thats not the intent here

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+            /* Decode message */
+            if (messages) {
+                ObjectMapper mapper = new ObjectMapper()
+                HookRequest request = mapper.readValue(messages.get(1), HookRequest.class)
```
Comment (GiriDandu) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: If you more than one message is returned in the list, what happens to the rest of messages?

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    void pop(Closure action) {
+        if (action == null) {
+            throw new Exception("Must provide a Closure to RedisQueue.pop()")
+        }
+
+        if (!this.started) {
+            throw new Exception("Queue must be started before this operation is invoked")
+        }
+
+        List<String> messages = null
```
Comment (GiriDandu) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: Can you make change this to List<String> messages = []

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+            /* Decode message */
+            if (messages) {
+                ObjectMapper mapper = new ObjectMapper()
+                HookRequest request = mapper.readValue(messages.get(1), HookRequest.class)
+                action.call(request)
+            }
+        } catch (all) {
+            /* Put this back on the front of the queue */
+            if (messages) {
```
Comment (GiriDandu) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: Check the messages.size()

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+     *
+     * If the request cannot be inserted, this method will return false,
+     * otherwise true.
+     */
+    Boolean push(HookRequest request) {
+        if (!this.started) {
+            throw new Exception("Queue must be started before this operation is invoked")
+        }
+
+        request.id = ++nextId
```
Comment (GiriDandu) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: Given whoas is a library, can this be an issue when two threads call this function? If it not thread safe, we should say that in the comment

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        try {
+            redisclient = redisClientFactory.getJedis()
+            ret = redisclient.rpush("queue", jsonPayload)
+        } catch (Exception e) {
+        } finally {
+            if (redisclient != null) {
+                redisClientFactory.returnJedis(redisclient)
+            }
+        }
+        return ret == 1;
```
Comment (GiriDandu) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: Cn rpush return value be > 1. For example below. we are specifically testing for "1". See below

edis> RPUSH mylist "hello"
(integer) 1
redis> RPUSH mylist "world"
(integer) 2

File: src/test/groovy/com/github/lookout/whoas/RedisQueueSpec.groovy [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    def "push() should put onto the internal queue"() {
+        given:
+        RedisQueue queue = new RedisQueue()
+        Jedis redisClient = new MockJedis("test");
+        queue.redisClientFactory.metaClass.getJedis = {redisClient}
+        redisClient.metaClass.rpush = {String key, String payload -> redisClient.lpush(key, payload)}
+
+        when:
+        queue.start()
+        queue.push(new HookRequest())
```
Comment (GiriDandu) [inline src/test/groovy/com/github/lookout/whoas/RedisQueueSpec.groovy]: Check the boolean return value

File: build.gradle [hunk 14]
```diff
@@ -35,7 +35,9 @@ dependencies {
     compile 'com.fasterxml.jackson.core:jackson-databind:2.3.3+'
     /* Needed for better time management/sanity */
     compile 'joda-time:joda-time:2.6+'
-
+    /* redis client */
+    compile 'redis.clients:jedis:2.6+'
+    compile 'com.fiftyonred:mock-jedis:0.4.0'
```
Comment (rtyler) [inline build.gradle]: This will cause both of these to be listed as transitive dependencies of whoas as a library which I do not believe is what you intended to do since it's not ideal behavior.

For integration testing these should both be in the `testCompile` configuration instead of `compile`

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 7 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: Its a list, returned by blpop

File: src/test/groovy/com/github/lookout/whoas/RedisQueueSpec.groovy (hunk 6 above)
Comment (maheshkelkar) [inline src/test/groovy/com/github/lookout/whoas/RedisQueueSpec.groovy]: done

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 8 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: index-0 has key
index-1 has value

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 9 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: I tried that, it still shows up as null

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 10 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: This checks if not empty & if not null; usually redis returns both the key & value; so I thought additional check is unnecessary

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 11 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: The push is only called by 1 thread. But, I can take it out if this is contentions. This is for debugging only

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 12 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: In this case we are pushing only 1 messages, then it will be either 0 or 1

File: src/test/groovy/com/github/lookout/whoas/RedisQueueSpec.groovy (hunk 13 above)
Comment (maheshkelkar) [inline src/test/groovy/com/github/lookout/whoas/RedisQueueSpec.groovy]: We dont have to, because getSize() will "verify" thats very thing.

File: src/main/groovy/com/github/lookout/whoas/AbstractHookQueue.groovy (hunk 2 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/AbstractHookQueue.groovy]: I think this is reasonable to put inside of the `AbstractHookQueue` class since just about every other queueing implementation will require a connection setup of some form or fashion

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy [hunk 15]
```diff
@@ -0,0 +1,62 @@
+package com.github.lookout.whoas
+
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.JedisPool;
+import redis.clients.jedis.JedisPoolConfig;
+
+public class RedisClientFactory {
+    private static final RedisClientFactory instance = new RedisClientFactory();
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: Semicolons are unnecessary in Groovy

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    private static JedisPool pool;
+    private RedisClientFactory() {}
+    public final static RedisClientFactory getInstance() {
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: This is unnecessary, just make the `instance` declaration above public, i.e.

```groovy
static final RedisClientFactory instance = new RedisClientFactory()
```

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy [hunk 17]
```diff
@@ -0,0 +1,62 @@
+package com.github.lookout.whoas
+
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.JedisPool;
+import redis.clients.jedis.JedisPoolConfig;
+
+public class RedisClientFactory {
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: This needs lots of Groovydoc explaining what this is here for and why

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy [hunk 18]
```diff
@@ -0,0 +1,62 @@
+package com.github.lookout.whoas
+
+import redis.clients.jedis.Jedis;
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: Semicolons!

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 12 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: I have fixed it to ret > 1

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+import redis.clients.jedis.JedisPoolConfig;
+
+public class RedisClientFactory {
+    private static final RedisClientFactory instance = new RedisClientFactory();
+    private static JedisPool pool;
+    private RedisClientFactory() {}
+    public final static RedisClientFactory getInstance() {
+        return instance;
+    }
+    public void start(String hostname, Integer port) {
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: `public` is not needed in Groovy, and this should have some Groovydocs on it

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+        /* Set JedisPoolConfig */
+        JedisPoolConfig poolConfig = new JedisPoolConfig();
+
+        /* Set max active connections to Redis instance */
+        poolConfig.setMaxTotal(10);
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: where/why is this constant of `10` here?

File: build.gradle (hunk 14 above)
Comment (maheshkelkar) [inline build.gradle]: I will move the mock-jedis to the testCompile. However we need "jedis" the redis client for communication with redis server

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+
+        /* Maximum number of connections to test in each idle check */
+        poolConfig.setNumTestsPerEvictionRun(10);
+
+        /* Idle connection checking period */
+        poolConfig.setTimeBetweenEvictionRunsMillis(60000);
+
+        /* Create the jedisPool */
+        pool = new JedisPool(poolConfig, hostname, port);
+    }
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: There's a lot of configuration options that are hard-coded that shouldn't be hard-coded here. The queue type should be able to take this options and use them with sane defaults

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import redis.clients.jedis.Jedis;
+
+/**
+ * A redis queue that offers distributed and persistent queue
+ */
+class RedisQueue extends AbstractHookQueue {
+    private final RedisClientFactory redisClientFactory
+    private static nextId = 0
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: What's nextId?

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+    private String hostname
+    private Integer port
+
+    /**
+     * Create the RedisQueue with valid config
+     */
+    RedisQueue(WhoasQueueConfig queueConfig) {
+        redisClientFactory = RedisClientFactory.getInstance()
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: `.getInstance()` -> `.instance`

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+    Long getSize() {
+        if (!this.started) {
+            throw new Exception("Queue must be started before this operation is invoked")
+        }
+        Jedis redisclient = null
+        Long queueSize = 0
+        try {
+            redisclient = redisClientFactory.getJedis()
+            queueSize = redisclient.llen("queue")
+        } catch (all) {
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: This empty catch all block doesn't make sense to me

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy [hunk 25]
```diff
@@ -0,0 +1,138 @@
...
+     */
+    Long getSize() {
+        if (!this.started) {
+            throw new Exception("Queue must be started before this operation is invoked")
+        }
+        Jedis redisclient = null
+        Long queueSize = 0
+        try {
+            redisclient = redisClientFactory.getJedis()
+            queueSize = redisclient.llen("queue")
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: The queue name of "queue" seems like something that should be configured

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy [hunk 26]
```diff
@@ -0,0 +1,138 @@
...
+        } catch (all) {
+        } finally {
+            if (redisclient != null) {
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: This is just a suggestion, but if you make `returnJedis` accept and handle a `null` then this line *could* be:

```groovy
redisClientFactory?.returnJedis(redisclient)
```

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+    public void stop() {
+        pool.destroy();
+    }
+    public Jedis getJedis() {
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: I suggest changing this "get" and "return" API to use a Closure instead, e.g.

```groovy
Object withRedis(Closure closure) {
    Jedis client = pool.resource
    try {
        return closure.call(client)
    }
    finally {
        client?.close()
    }
}
```

This way the access pattern below could become:

```groovy
queueSize = factory.withRedis { r -> r.llen('myqueue') }
```

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy (hunk 17 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: I should not that this object isn't so much a factory as it is a wrapper around the JedisPool

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy (hunk 15 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: copy+paste error; I will take them ";" out

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 7 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: [BLPOP](http://redis.io/commands/blpop) should only be returning a single element

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 11 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: Agreed that this isn't thread-safe and it's not very useful either, we should just be using a UUID generator function here. I'm pretty sure there's something in the Java standard library for this

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 12 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: Any non-zero return value is successful

File: src/main/groovy/com/github/lookout/whoas/WhoasQueueConfig.groovy [hunk 28]
```diff
@@ -0,0 +1,34 @@
+package com.github.lookout.whoas
+
+import com.fasterxml.jackson.annotation.JsonProperty
+
+/**
+ *  Whoas Queue Configuration
+ */
+public class WhoasQueueConfig {
```
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/WhoasQueueConfig.groovy]: With the configuration change to a Map for describing the queue type and the options to be passed into the queue constructor, I no longer understand the utility of this class

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 22 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: debugging

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 24 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: Its catching all the exception thrown by jedis; RedisQueue will return an error to the whoas client. 

If general approach is throw an exception, I can come up with a Whoas specific excpetions

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy (hunk 21 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: Agreed. there is a lot of configuration and hardcoding. However, there is no value in exposing a lot of these options until this code matures enough. I have borrowed this code from the web, I kept the parameters which made sense

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy (hunk 20 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: This puts a max limit on number of Jedis client instantiaions. It seems sane number to begin with. In reality I dont think we need to more than 3/4 i.e. number of parallel operations.

IMO we can put a default right now and make it configurable depending upon how this matures

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy (hunk 17 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: Yes its  a wrapper, but it also dispenses the Jedis instances. Hence called it a factory. I will add a javadocs

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 25 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: I just hardcoded for the timebeing, because I am not sure if we will have one or multiple queues in future. We havent discussed that part yet. 

E.g. Should we use "lpoprpush mainq tempq" instead of "lpop mainq" and keep that message in the redis until its consumed. How do we manage these tempq messages etc.

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy (hunk 17 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: It's not "giving those instances away" though, it's only providing a borrowing agreement on them. Naming this Factory I find counter-intuitive since all the other factories we use are *building* instances for the consumer, whereas this abstraction is only lending pre-configured instances.

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy (hunk 27 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: Does "Object" work for all dataTypes? I had sent 2 emails about this yesterday and you answered them now :)

As I understand, pool.resource may throw an exception, so that has to be inside "try". I dont think close throws anything.

What if closure.call throws an exception, do you think it will reach the caller of withRedis?

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy (hunk 20 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: Defaulting is fine, but `start()` still isn't the right place to codify all these defaults. Even if we don't plumb together the configuration right now, I'd like to see these properties overwritable by consumers of this class

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 7 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: BLPOP? yes
Jedis client API returns a list: public List<String> blpop(final int timeout, final String... keys) {

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 11 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: I will take it out; we can add a uuid or something to put identifier on the hook request

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 24 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: This kind of code is really not okay, like, ever. Catching swallowing exceptions means that nobody will ever know what happens on errors here. If the `llen()` call throws we won't know, we'll just return zero, which incorrect. I would rather let all our exceptions from Jedis be thrown further up. Catching and wrapping in Whoas speciifc exceptions I don't think would provide any value either, but just serve as a middle-man exception class which hides information from a developer

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 8 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: That's nuts, but if that's what Jedis is doing then please document this in code comments

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy (hunk 17 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: I can argue otherwise, but you have point. I will rename it to wrapper

File: src/main/groovy/com/github/lookout/whoas/WhoasQueueConfig.groovy (hunk 28 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/WhoasQueueConfig.groovy]: We didnt change it to a map. The config in the deploydb will be as follows. Its a struct.
queue:
  type: "class-name"
  hostname: "localhost"
  port: 6329

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy (hunk 17 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: See my comments in Gitter, i'm positing that this class simply is unnecessary wrapping around `JedisPool` and this logic should be internal to the `RedisQueue` class

File: src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy (hunk 27 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisClientFactory.groovy]: > Does "Object" work for all dataTypes? I had sent 2 emails about this yesterday and you answered them now :)

It does, everything is an Object (supposedly) in Java/Groovy land :package: 

> As I understand, pool.resource may throw an exception, so that has to be inside "try". I dont think close throws anything.

[These javadocs](http://javadox.com/redis.clients/jedis/2.2.0/redis/clients/util/Pool.html#getResource%28%29) don't indicate that `getResource` should raise, if it does, that's okay because nothing will have been checked out that would need to be closed


> What if closure.call throws an exception, do you think it will reach the caller of withRedis?

It will

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 22 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: Comment it as such please, I think this is reasonable to have on instances (not static)

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 25 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: I suggest adding "key" to the configuration for RedisQueue which would be used here, such that a YAML would look like:

```yaml
queue:
  type: com.github.lookout.whoas.RedisQueue
  key: "my-queue"
  hostname: "localhost"
  port: 1337
```

We did not talk about multiple queues or anything like that, but I think we should definitely just support one for now; hard for me to imagine multiple queue support being a reasonable request of such a simple little library

File: src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy (hunk 24 above)
Comment (maheshkelkar) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: Jedis when not connected to Redis server, may throw SocketTimeoutException (which can happen in normal scenario) or some other exception if not enought Jedis available in the pool. 

In that case exposing the exception may cause webhook invocation from deploydb to fail (because deployDb doesnt catch that. 

Yes, catch all is not sensible, but I thought we are playing with fire here (a lot of unknowns) and it seemed like conservative approach to begin with. But, I can dumb it down to certain exceptions and throw otherwise.

what do you say?

Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/RedisQueue.groovy]: I say, we should take the catch out for now and just raise up. Once #9 is completed then we can swallow these exceptions and return an invalid queueSize (-1) while still properly logging the full exception :)

File: src/main/groovy/com/github/lookout/whoas/WhoasQueueConfig.groovy (hunk 28 above)
Comment (rtyler) [inline src/main/groovy/com/github/lookout/whoas/WhoasQueueConfig.groovy]: Okay, this is contrary to what I thought we discussed gitter this morning, ping me when you're back around there and we can chat further

**Comment by rtyler** [comment]
This looks great, I'm going to rebase this for you and merge it. Thanks for dealing with all my code review bullshit :)

**Comment by rtyler** [comment]
Rebased in #20
